R/fonts.R

```diff
@@ -0,0 +1,444 @@
+
+  font_object(x, font_dep_face)
+}
+
+font_face_css <- function(x) {
+  props <- names(x)
+  font_prop <- !props %in% c("src", "unicode_range")
+  props[font_prop] <- paste0("font-", props[font_prop])
+  paste0(
+    "@font-face {\n",
+    paste0("  ", props, ": ", x, ";", collapse = "\n"),
+    "\n}"
+  )
+}
+
+font_object <- function(x, dep_func) {
+  stopifnot(is.function(dep_func))
+  # Dependency functions want to use unquoted family name
+  new_font_collection(
+    families = verify_single_unquoted_family(x$family),
+    # Produce dependency at render-time (i.e., tagFunction())
+    # so the context-aware caching dir has the proper context
+    html_deps = tagFunction(function() dep_func(x))
+  )
+}
+
+#' @rdname font_face
+#' @param ... a collection of `font_google()`, `font_link()`, `font_face()`, and/or character strings. Character strings must define a single `font-family`.
+#' @param default_flag whether or not to include a `!default` when converted to a Sass variable with [as_sass()].
+#' @export
+font_collection <- function(..., default_flag = TRUE) {
+  fonts <- list2(...)
+  families <- lapply(fonts, function(x) {
+    if (is_font_collection(x))
+      return(x$families)
+    if (is_string(x))
+      return(verify_single_unquoted_family(x))
+    stop(
+      "`font_collection()` expects a collection of `font_google()`, `font_link()`, `font_face()`, and/or character strings.",
+      call. = FALSE
+    )
+  })
+  families <- unlist(families, recursive = FALSE, use.names = FALSE)
+  deps <- lapply(fonts, function(x) {
+    if (is_font_collection(x)) x$html_deps
+  })
+  new_font_collection(
+    families = families,
+    html_deps = unlist(deps, recursive = FALSE, use.names = FALSE),
+    default_flag = isTRUE(default_flag)
+  )
+}
+
+new_font_collection <- function(families, html_deps, default_flag = TRUE) {
+  add_class(
+    list(
+      families = families,
+      html_deps = html_deps,
+      default_flag = default_flag
+    ),
+    "font_collection"
+  )
+}
+
+#' @rdname font_face
+#' @param x test whether `x` is a `font_collection()`, `font_google()`, `font_link()`, or `font_face()` object.
+#' @export
+is_font_collection <- function(x) {
+  inherits(x, "font_collection")
+}
+
+# Only to be used when we know x is meant to be a single font family
+verify_single_unquoted_family <- function(x) {
+  if (!is_string(x)) {
+    stop("Font family definitions must be a character string (length 1).", call. = FALSE)
+  }
+  has_comma <- grepl(",", x)
+  quoted <- grepl("'", x) || grepl('"', x)
+  if (has_comma || quoted) {
+    stop(
+      "This family definition contains ",
+      if (has_comma) "a comma: " else "quotes: ", x, ". ",
+      "Font family definitions must define a single (unquoted) family name. ",
+      call. = FALSE
+    )
+  }
+  x
+}
+
+font_dep_name <- function(x) {
+  paste0(class(x)[[1]], sub("\\s*", "_", tolower(x$family)))
+}
+
+#' @import htmltools
+font_dep_face <- function(x) {
+  src_dir <- tempfile()
```

Comment:
At least add a comment here about (someday?) doing something a kin to `output_template` (or memoise?) to avoid un-neccessary file creation?